privacyidea/lib/tokens/pushtoken.py

```diff
@@ -0,0 +1,578 @@
+                        u"The message could not be sent.".format(self.token.serial,
+                                                                 PUSH_ACTION.FIREBASE_CONFIG))
+
+        validity = int(get_from_config('DefaultChallengeValidityTime', 120))
+        tokentype = self.get_tokentype().lower()
+        # Maybe there is a PushChallengeValidityTime...
+        lookup_for = tokentype.capitalize() + 'ChallengeValidityTime'
+        validity = int(get_from_config(lookup_for, validity))
+
+        # Create the challenge in the database
+        db_challenge = Challenge(self.token.serial,
+                                 transaction_id=transactionid,
+                                 challenge=challenge,
+                                 data=data,
+                                 session=options.get("session"),
+                                 validitytime=validity)
+        db_challenge.save()
+        self.challenge_janitor()
+        return True, message, db_challenge.transaction_id, attributes
+
+    @check_token_locked
+    def check_challenge_response(self, user=None, passw=None, options=None):
+        """
+        This function checks, if the challenge for the given transaction_id
+        was marked as answered correctly.
+        For this we check the otp_status of the challenge with the
+        transaction_id in the database.
+
+        We do not care about the password
+
+        :param user: the requesting user
+        :type user: User object
+        :param passw: the password (pin+otp)
+        :type passw: string
+        :param options: additional arguments from the request, which could
+                        be token specific. Usually "transaction_id"
+        :type options: dict
+        :return: return otp_counter. If -1, challenge does not match
+        :rtype: int
+        """
+        options = options or {}
+        otp_counter = -1
+
+        # fetch the transaction_id
+        transaction_id = options.get('transaction_id')
+        if transaction_id is None:
+            transaction_id = options.get('state')
+
+        # get the challenges for this transaction ID
+        if transaction_id is not None:
+            challengeobject_list = get_challenges(serial=self.token.serial,
+                                                  transaction_id=transaction_id)
+
+            for challengeobject in challengeobject_list:
+                # check if we are still in time.
+                if challengeobject.is_valid():
+                    _, status = challengeobject.get_otp_status()
+                    if status is True:
+                        # create a positive response
+                        otp_counter = 1
+                        # delete the challenge, should we really delete the challenge? If we do so, the information
+                        # about the successful authentication could be fetched only once!
+                        # challengeobject.delete()
```

Comment:
Hm, I would have thought we do have to delete the challenge. :-) But I don't have a full understanding of the protocol yet.

Reply:
The thing is, that we keep the challenge for a while, so that we can fetch the information several times.
The old expired challenge gets deleted by the challenge janitor, when a new challenge is created for this very token.
https://github.com/privacyidea/privacyidea/blob/1342-pushtoken/privacyidea/lib/tokens/pushtoken.py#L530